### PR TITLE
SI-8930 - Vector updated, +:, and :+ slow when typed as Seq[A]

### PR DESCRIPTION
**Pull request by Ichoran**
Vector was intercepting only the IndexedSeq CanBuildFrom to quickly generate new vectors.  Now it intercepts immutable.Seq and collection.Seq as well.

There are other possibilities (collection.IndexedSeq), but they will probably arise rarely, and to avoid an absurdly long set of checks we would need a marker trait (that is not binary compatible).

### CODE REVIEW

**Comment by adriaanm** [comment]
Since the 2.11.6 deadline is extremely nigh, I've tentatively postponed this PR (along with other recent ones) to 2.11.7. I'll make a pass through reviewed 2.11.7 PRs on Friday and move them back to 2.11.6 where feasible.

In the mean time, feel free to let me know if you'd like some help getting this reviewed by then.

**Comment by Ichoran** [comment]
Rebased just to clear this out of my queue of things-to-think-about-for-now.

**Comment by okumin** [comment]
Like List, can we solve this problem by testing whether bf is Vector.GenericCanBuildFrom[_] or not?
Or are there corner cases?

https://github.com/scala/scala/blob/2.11.x/src/library/scala/collection/immutable/List.scala#L199

**Comment by adriaanm** [comment]
I wonder if it wouldn't be faster to use a marker trait and an instanceof test instead of multiple `eq`s. Pushing to 2.11.8 while we ponder this.

**Comment by Ichoran** [comment]
@adriaanm - We can't add a marker trait in 2.11 anyway.  So I'm not sure there's much to ponder, unless you mean to wait to 2.12 to fix this?

**Comment by adriaanm** [comment]
Right you are, moving back to 2.11.7. (Was in bulk mode for pruning PRs due to imminent 2.11.7 release.)

**Comment by adriaanm** [comment]
I'm still a bit uncomfortable with the copy/pasting of all these eq tests. Can we factor them out in an inlined private method? We're in a class, so inlining should work even in the current optimizer.

**Comment by Ichoran** [comment]
@adriaanm - Good point.  Updated.

File: src/library/scala/collection/immutable/Vector.scala [hunk 1]
```diff
@@ -211,7 +217,8 @@ override def companion: GenericCompanion[Vector] = Vector
 
   // concat (suboptimal but avoids worst performance gotchas)
   override def ++[B >: A, That](that: GenTraversableOnce[B])(implicit bf: CanBuildFrom[Vector[A], B, That]): That = {
-    if (bf eq IndexedSeq.ReusableCBF) {
+    if ((bf eq IndexedSeq.ReusableCBF) || (bf eq collection.immutable.Seq.ReusableCBF) || (bf eq collection.Seq.ReusableCBF)) {
```
Comment (adriaanm) [inline src/library/scala/collection/immutable/Vector.scala]: Could this use `isDefaultCBF` as well?

Comment (Ichoran) [inline src/library/scala/collection/immutable/Vector.scala]: Whoops.  Yes, it could.

**Comment by okumin** [comment]
Can we use `Vector.GenericCanBuildFrom[_]` as a marker trait?
It probably matches Vector's GenericCanBuildFrom, IndexedSeq's GenericCanBuildFrom, Seq's GenericCanBuildFrom and GenericCanBuildFroms of the other super classes of Vector.

I'm sorry if I was wrong.

**Comment by Ichoran** [comment]
@okumin - I recall deciding before that this wouldn't work, but I forget why--but note that there is a lot of casting to single instances going on in the various CanBuildFroms.  Do you have a solid idea of why it would work (so I don't have to go digging through all the sources again)?

**Comment by adriaanm** [comment]
Thank you for your suggestion, @okumin! I think it's very worthwhile to explore it. Unfortunately, the deadline for 2.11.7 is today, and I'd like to get this PR merged. We will be very happy to receive another PR that improves on the current design!

**Pull request merged**